### ISSUE TITLE
General questions

### ISSUE THREAD

**Issue by jazzl0ver** [goadapp/goad#28]
Hello,

Can you please answer the following questions:

1 - Am I correct that TotErrors means connection errors happened during tcp handshake?

2 - I have nginx 1.8 running on c4.4xlarge Amazon Linux instance (16 CPUs, bunch of memory), kernel is tuned like this:

```
net.ipv4.tcp_syncookies = 0
net.ipv4.ip_local_port_range=4000 65535
net.ipv4.tcp_fin_timeout=2
net.ipv4.tcp_tw_recycle=1
net.ipv4.tcp_tw_reuse=1
net.core.somaxconn=1024
net.core.netdev_max_backlog=65535
net.ipv4.tcp_max_syn_backlog=65535
net.core.rmem_max = 16777216
net.core.wmem_max = 16777216
net.core.rmem_default = 8388608
net.core.wmem_default = 8388608
net.ipv4.tcp_rmem = 4096 8388608 16777216
net.ipv4.tcp_wmem = 4096 8388608 16777216
fs.file-max = 1000000
fs.nr_open = 1000000
```

nginx meaningful options are:

```
worker_processes 16;
worker_cpu_affinity 0001 0010 0100 1000 01000 10000 010000 100000 0100000 1000000 01000000 10000000 010000000 100000000 0100000000 1000000000;

events {
    worker_connections 655350;
}
```

goad is started with:

```
$ build/linux/x86-64/goad -c 25000 -n 2000000 -r us-east-1 -t 30 -u "http://verygoodhost.dom/"
```

The results differ from time to time:

```
   TotReqs   TotBytes    AvgTime   AvgReq/s  AvgKbps/s
   2000000      32 MB     0.023s  285493.59    4363.29
   Slowest    Fastest   Timeouts  TotErrors
   15.040s     0.000s          0      27467
HTTPStatus   Requests
       200    1972533
```

```
   TotReqs   TotBytes    AvgTime   AvgReq/s  AvgKbps/s
   2000000      30 MB     0.023s   40320.49     584.64
   Slowest    Fastest   Timeouts  TotErrors
   15.057s     0.000s          0      97712
HTTPStatus   Requests
       200    1902288
```

```
   TotReqs   TotBytes    AvgTime   AvgReq/s  AvgKbps/s
   1933334      29 MB     0.023s    8653.97     120.74
   Slowest    Fastest   Timeouts  TotErrors
    7.218s     0.000s          0     129231
HTTPStatus   Requests
       200    1804103
```

```
   TotReqs   TotBytes    AvgTime   AvgReq/s  AvgKbps/s
   2000000      30 MB     0.023s  571338.56    7865.94
   Slowest    Fastest   Timeouts  TotErrors
   15.075s     0.000s          0      97252
HTTPStatus   Requests
       200    1902748
```

Can you please explain why that might happen?

3 - When I set concurrency to 25000 I expect there will be 25000 opened sockets on the target server. But it seems to be a wrong assumption:

```
# netstat -an|wc -l
6352
# netstat -an|wc -l
3795
# netstat -an|wc -l
2104
# netstat -an|wc -l
3620
# netstat -an|wc -l
3832
# netstat -an|wc -l
4477
# netstat -an|wc -l
2654
# netstat -an|wc -l
742
```

Netstat ran almost every second or two while the test was running. Nginx's  internal status page also does not show 25000 connections. Why?

4 - Do you have any recommendations for tuning the system in a better way?

**Comment by danielmotaleite** [comment]
1- yes, that one is connection errors.
If after connecting, the server returns 503, it will not be counted here but goes to the HTTPStatus table below

2- aws ec2 are "shared" resources...  sometimes they are faster, other slower... Disk IO and network IO sometime can vary a lot.
you can try a similar test on real hardware with the same parameters and how it behaves.
Also, sometime is better to have several smaller machines than just one big one (all eggs in single basket vs several baskets), so try also that and see if you get better results or more consistent ones
Finally, you might be also hitting lambda limits, try to use more zones to spread the lambda requests and avoid being throttled. I also have a pull request that tries to spread better the requests, it might help if you are hitting limits

3- your result show that you have 0s response time, so probably your connections are being replied so fast that netstat can only see those active connections, the other ones are being open or are already closed. you can try to create a iptables rule that logs all new connections and you can check that way how many of then are open.
Try to request a small static file (i use favicon.ico) and then a bigger file (watch out for volume of traffic) and you should see a difference in the number of open ports. Also, try to play with the keepalive values,it will keep the connection open longer.

Finally, have you any other system in between? like a CDN, load balancer? those can open less backend connections and reuse then, so less connections for the server to manage.

4- it is always hard to give this kind of tips without knowing the systems, but nginx is fast and looking at your machine, it looks too big...  unless it is a loadbalancer, it seems weird.

Try to do several smaller goad tests and start to decrease the number of workers and repeat the tests and see how it behaves.... i suspect it will not make huge difference... and they start to increase the concurrency and again play with the workers.

**Comment by jazzl0ver** [comment]
Thank you very much!

**Comment by danielmotaleite** [comment]
> 1- yes, that one is connection errors.
> If after connecting, the server returns 503, it will not be counted here but goes to the HTTPStatus table below

Sorry... I'm wrong where... TotErrors looks like it is  timeout+connection errors+http 5xx and probably 4xx. I just finished a bunch of tests and i got concistent TotErrors = 503 errors, so yes, they are related. I also got results in the past with TotErrors was high, but very small 503, so it also accounts connections errors (should be more common in https)

Also notice the TotReqs, it might be below what you requested, if the lambda for some reason didn't run the request. So the missing requests are not counted in any place
